File: src/srcom/verified.py
#!/usr/bin/env python3.9

"""
This program gets the number of runs verified and rejected by a given user (argv[1]) and optionally
limits the count to runs from 1 or 2 given games (argv[2] and argv[3]).
"""

import json
from collections import defaultdict
from datetime import datetime
from fcntl import LOCK_EX, LOCK_NB, LOCK_UN, flock
from io import TextIOWrapper
from itertools import count
from os.path import exists
from sys import argv, exit
from time import sleep

import requests
from requests import Session
from utils import *

TWO_HOURS = 7200
USAGE = (
	"Usage: `+verified [PLAYER NAME] [GAME (Optional)] [GAME (Optional)]`\n"
	+ "Example: `+verified AnInternetTroll mkw mkwextracategories`"
)


class xopen:
	"""
	A class that can be used similarly to open() that locks the file before reading or writing
	and releases the lock when done. This is useful to avoid things like +vlb from spawning
	multiple +verified processes that all try to access the cache at the same time.
	"""

	def __init__(self, filename: str, mode: str) -> None:
		self.filename = filename
		self.mode = mode
		self.file: TextIOWrapper = None

	def __enter__(self) -> TextIOWrapper:
		while True:
			try:
				self.file = open(self.filename, self.mode, encoding="utf-8")
				flock(self.file, LOCK_EX | LOCK_NB)
			except BlockingIOError:
				sleep(0.05)
			except OSError as e:
				error_and_die(e)
			else:
				return self.file

	def __exit__(self, _, __, ___) -> None:
		flock(self.file, LOCK_UN)
		self.file.close()


def xapi_get(session: Session, uri: str, params: Optional[dict[str, Any]] = {}) -> dict:
	"""
	A copy of what we have in utils.py but adapted for `requests.Session()`.
	"""
	while True:
		try:
			with session.get(uri, params=params) as r:
				if r.ok:
					return r.json()
				if r.status_code == RATE_LIMIT:
					sleep(5)
				else:
					try:
						error_and_die(r.json()["message"])
					except JSONDecodeError:
						error_and_die("The site is probably down go complain to ELO")
		except ConnectionError:
			sleep(2)


def check_cache_exists() -> None:
	"""
	A simple function to check if the cache file exists, and if it doesn't, to create it.
	"""
	if not exists(f"{CACHEDIR}/verified.json"):
		with xopen(f"{CACHEDIR}/verified.json", "w") as f:
			f.write("{}")


def write_to_cache(uid: str, totals: defaultdict) -> None:
	"""
	Writes the examination counts stored in the `totals` default dict to the cache.
	"""
	with xopen(f"{CACHEDIR}/verified.json", "r+") as f:
		data = json.load(f)

		if uid not in data:
			data[uid] = {}
		for game in totals:
			data[uid][game] = totals[game]

		data[uid]["last_updated"] = datetime.timestamp(datetime.utcnow())

		f.seek(0)
		json.dump(data, f, indent=8)
		f.truncate()


def fetch_runs(session: Session, uid: str, offset: int, totals: defaultdict) -> int:
	"""
	Takes a requests session and makes a request, adding the appropriate examination counts to
	the totals default dict. In the case of rate limiting it sleeps for 5 seconds and tries
	again.
	"""
	ret = 0
	while True:
		data = xapi_get(
			session, f"{API}/runs", params={"examiner": uid, "max": 200, "offset": offset}
		)

		for run in data["data"]:
			totals[run["game"]] += 1

		ret += data["pagination"]["size"]
		return ret


def make_requests(uid: str, gids: list[str]) -> int:
	"""
	Takes a user ID and a list of game IDs. Requests are made to get all the runs examined by
	the specified user and the totals for each game in the cache are updated. The number of
	examined runs for the game(s) specified by the user are then returned (or the total if none
	specified).
	"""
	totals: defaultdict = defaultdict(int)
	with requests.Session() as session:
		for i in count(0, 2000):
			tmp = 0
			for j in range(i, i + 2000, 200):
				tmp += fetch_runs(session, uid, j, totals)

			totals["total"] += tmp
			if tmp < 2000:
				break

	write_to_cache(uid, totals)
	if not gids:
		return totals["total"]
	return sum(totals[i] for i in gids)


def examined(uid: str, gids: list[str]) -> int:
	"""
	Gets the number of runs examined by the given user for the given games. If no games are
	specified then it gets all the users examined runs.
	"""
	check_cache_exists()
	with xopen(f"{CACHEDIR}/verified.json", "r") as f:
		data = json.load(f)

	try:
		then = data[uid]["last_updated"]
		now = datetime.timestamp(datetime.utcnow())
	except KeyError:  # User hasn't been added to cache yet
		return make_requests(uid, gids)

	if now - then > TWO_HOURS:
		return make_requests(uid, gids)

	if gids:
		try:
			return sum(data[uid][gid] for gid in gids)
		except KeyError:  # Counts for a specific game havent been added to cache yet
			return make_requests(uid, gids)
	return data[uid]["total"]


def getgids(games: list[str]) -> list[str]:
	"""
	Returns a list of game IDs for each game specified by the user.
	"""
	gids: list[str] = []
	for game in games:
		_, gid = getgame(game)
		gids.append(gid)

	return gids


def main() -> int:
	if not (1 < (argc := len(argv)) < 5):
		usage(USAGE)

	# Avoid the same game being passed twice
	if argc == 4 and argv[2] == argv[3]:
		argv.pop()

	uid = getuid(argv[1])
	gids = getgids(argv[2:])
	total = examined(uid, gids)

	print(f"Verified: {total}")
	return EXIT_SUCCESS


if __name__ == "__main__":
	exit(main())

File: src/srcom/verified.c
/*
 * This program gets the number of runs verified and rejected by a given user
 * (argv[1]) and optionally limits the count to runs from 1 or 2 given games
 * (argv[2] and argv[3]).
 */

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>

#include <curl/curl.h>

#include "defines.h"
#include "srcom/utils.h"
#include "srcom/verified.h"

static void
usage(void)
{
	fputs("Usage: `+verified [PLAYER NAME] [GAME (Optional)] [GAME (Optional)]`\n"
	      "Example: `+verified AnInternetTroll mkw mkwextracategories`\n",
	      stderr);
	exit(EXIT_FAILURE);
}

static unsigned int
count_examined(bool *done, string_t *json)
{
	char *ptr = json->ptr;
	unsigned int count = 0, tmp;

	while ((ptr = strstr(ptr + KEY_LEN, SIZE_KEY)) != NULL) {
		sscanf(ptr, SIZE_KEY "%u", &tmp);

		if (tmp < MAX_RECV)
			*done = true;
		count += tmp;
	}

	return count;
}

static void
perform_requests(char *uri_base, unsigned int offset, string_t *json)
{
	char uri[URIBUF];
	int running = 0, numfds;
	CURLcode mc;
	CURL *handles[REQUEST_COUNT];
	CURLM *mhandle;

	if ((mhandle = curl_multi_init()) == NULL) {
		fputs("Error: Unable to initialize curl multi handle.", stderr);
		exit(EXIT_FAILURE);
	}

	for (int i = 0; i < REQUEST_COUNT; i++) {
		if ((handles[i] = curl_easy_init()) == NULL) {
			fputs("Error: Unable to initialize curl handle.\n",
			      stderr);
			exit(EXIT_FAILURE);
		}

		sprintf(uri, "%s%u", uri_base, offset);
		offset += MAX_RECV;

		/* Load the contents of the API request to `json`. */
		curl_easy_setopt(handles[i], CURLOPT_URL, uri);
		curl_easy_setopt(handles[i], CURLOPT_WRITEFUNCTION,
		                 write_callback);
		curl_easy_setopt(handles[i], CURLOPT_WRITEDATA, json);
		curl_easy_setopt(handles[i], CURLOPT_FAILONERROR, 1L);

		curl_multi_add_handle(mhandle, handles[i]);
	}

	do {
		mc = curl_multi_perform(mhandle, &running);
		if (mc == CURLM_OK)
			mc = curl_multi_poll(mhandle, NULL, 0, 1000, &numfds);
		else {
			fprintf(stderr,
			        "Error: curl_multi_perform failed, code %d.\n",
			        mc);
			exit(EXIT_FAILURE);
		}

		if (mc != CURLM_OK) {
			fprintf(stderr, "Error: curl_multi failed, code %d.\n",
			        mc);
			exit(EXIT_FAILURE);
		}
	} while (running);

	/* Cleanup */
	curl_multi_cleanup(mhandle);
	for (int i = 0; i < REQUEST_COUNT; i++)
		curl_easy_cleanup(handles[i]);
}

static unsigned int
get_examined(const char *uid, const char *gname)
{
	bool done = false;
	char uri_base[URIBUF];
	unsigned int examined = 0, offset = 0;
	string_t json;
	struct game_t *game = NULL;

	if (gname && (game = get_game(gname)) == NULL) {
		fprintf(stderr,
		        "Error: Game with abbreviation '%s' not found.\n",
		        gname);
		exit(EXIT_FAILURE);
	}

	sprintf(uri_base, API "/runs?examiner=%s&game=%s&max=" STR(MAX_RECV) "&offset=",
	         uid, game ? game->id : "");

	/* Every loop `json` is cleared */
	while (!done) {
		init_string(&json);

		perform_requests(uri_base, offset, &json);
		examined += count_examined(&done, &json);
		offset += MAX_RECV * REQUEST_COUNT;

		free(json.ptr);
	}

	return examined;
}

int
main(int argc, char **argv)
{
	if (argc < 2)
		usage();

	char *uid;
	unsigned int examined;

	if ((uid = get_uid(argv[1])) == NULL) {
		fprintf(stderr, "Error: User with username '%s' not found.\n",
		        argv[1]);
		exit(EXIT_FAILURE);
	}

	examined = get_examined(uid, argv[2]);
	/* Check for the case where the same game is given twice */
	if (argc > 3 && strcmp(argv[2], argv[3]) != 0)
		examined += get_examined(uid, argv[3]);

	printf("Verified: %u\n", examined);
	return EXIT_SUCCESS;
}
